File: dkplayer-java/src/main/java/xyz/doikki/videoplayer/render/TextureViewRender.java
package xyz.doikki.videoplayer.render;

import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.SurfaceTexture;
import android.util.AttributeSet;
import android.view.Surface;
import android.view.TextureView;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.lang.ref.WeakReference;

import xyz.doikki.videoplayer.AVPlayer;
import xyz.doikki.videoplayer.VideoViewManager;

/**
 * 经查看{@link TextureView}源码，发现在{@link #onDetachedFromWindow()}的时候，
 * 会先回调{@link TextureView.SurfaceTextureListener#onSurfaceTextureDestroyed}并释放{@link SurfaceTexture},
 * 在需要{@link SurfaceTexture}时会重新构建并回调{@link TextureView.SurfaceTextureListener#onSurfaceTextureAvailable}
 */
@SuppressLint("ViewConstructor")
public class TextureViewRender extends TextureView implements Render, TextureView.SurfaceTextureListener {

    private final RenderLayoutMeasure mMeasureHelper;

    @Nullable
    private WeakReference<AVPlayer> mPlayerRef;

    private SurfaceTexture mSurfaceTexture;
    private Surface mSurface;

    private SurfaceListener mSurfaceListener;

    public TextureViewRender(Context context) {
        super(context);
    }

    public TextureViewRender(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }

    public TextureViewRender(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    {
        mMeasureHelper = new RenderLayoutMeasure();
        setSurfaceTextureListener(this);
    }

    /**
     * 是否开启了渲染优化
     *
     * @return
     */
    private boolean isEnableRenderOptimization() {
        return VideoViewManager.isTextureRenderOptimization();
    }

    /**
     * 绑定播放器
     *
     * @param player
     */
    @Override
    public void attachToPlayer(@NonNull AVPlayer player) {
        try {
            if (mPlayerRef != null) {
                AVPlayer previousPlayer = mPlayerRef.get();
                if (previousPlayer != null && previousPlayer != player) {
                    //如果之前已绑定过播放器，并且与当前设置的播放器不相同，则将surface与之前的播放器解除绑定
                    previousPlayer.setSurface(null);
                    mPlayerRef = null;
                }
            }
            mPlayerRef = new WeakReference<>(player);
        } catch (Throwable e) {
            e.printStackTrace();
        }
    }

    @Override
    public void setVideoSize(int videoWidth, int videoHeight) {
        if (videoWidth > 0 && videoHeight > 0) {
            mMeasureHelper.setVideoSize(videoWidth, videoHeight);
            requestLayout();
        }
    }

    /*************START Render 实现逻辑***********************/

    @Override
    public View getView() {
        return this;
    }

    @Override
    public void setSurfaceListener(SurfaceListener listener) {
        this.mSurfaceListener = listener;
    }

    @Override
    public void screenshot(boolean highQuality, @NonNull ScreenShotCallback callback) {
        if (!isAvailable()) {
            callback.onScreenShotResult(null);
            return;
        }

        if (highQuality) {
            callback.onScreenShotResult(getBitmap());
        } else {
            callback.onScreenShotResult(getBitmap(Utils.createShotBitmap(this, false)));
        }
    }

    /*************END Render 实现逻辑***********************/

    /*************START TextureView.SurfaceTextureListener 实现逻辑***********************/

    @Override
    public void onSurfaceTextureAvailable(SurfaceTexture surfaceTexture, int width, int height) {
        if (isEnableRenderOptimization()) {
            //开启渲染优化
            if (mSurfaceTexture == null) {
                mSurfaceTexture = surfaceTexture;
                mSurface = new Surface(surfaceTexture);
                bindSurfaceToMediaPlayer(mSurface);
            } else {
                //在开启优化的情况下，使用最开始的那个渲染器
                setSurfaceTexture(mSurfaceTexture);
            }
        } else {
            mSurface = new Surface(surfaceTexture);
            bindSurfaceToMediaPlayer(mSurface);
        }
        notifySurfaceAvailable(mSurface, width, height);
    }

    @Override
    public boolean onSurfaceTextureDestroyed(SurfaceTexture surface) {
        //清空释放
        if (mSurfaceListener != null) {
            mSurfaceListener.onSurfaceDestroyed(mSurface);
        }

        if (isEnableRenderOptimization()) {
            //如果开启了渲染优化，那mSurfaceTexture通常情况不可能为null（在onSurfaceTextureAvailable初次回调的时候被赋值了），
            // 所以这里通常返回的是false，返回值false会告诉父类不要释放SurfaceTexture
            return (mSurfaceTexture == null);
        } else {
            return true;
        }
    }

    @Override
    public void onSurfaceTextureSizeChanged(SurfaceTexture surface, int width, int height) {
        if (mSurfaceListener != null) {
            mSurfaceListener.onSurfaceSizeChanged(mSurface, width, height);
        }
    }

    @Override
    public void onSurfaceTextureUpdated(SurfaceTexture surface) {
        if (mSurfaceListener != null) {
            mSurfaceListener.onSurfaceUpdated(mSurface);
        }
    }

    private void bindSurfaceToMediaPlayer(Surface surface) {
        if (mPlayerRef == null)
            return;
        AVPlayer player = mPlayerRef.get();
        if (player == null)
            return;
        player.setSurface(surface);
    }

    private void notifySurfaceAvailable(Surface surface, int width, int height) {
        if (mSurfaceListener != null) {
            mSurfaceListener.onSurfaceAvailable(surface);
        }
    }

    /*************END TextureView.SurfaceTextureListener 实现逻辑***********************/

    @Override
    public void setVideoRotation(int degree) {
        mMeasureHelper.setVideoRotationDegree(degree);
        setRotation(degree);
    }

    @Override
    public void setScaleType(int scaleType) {
        mMeasureHelper.setAspectRatioType(scaleType);
        requestLayout();
    }


    @Override
    public void release() {
        if (mSurface != null)
            mSurface.release();

        if (mSurfaceTexture != null)
            mSurfaceTexture.release();
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        mMeasureHelper.doMeasure(widthMeasureSpec, heightMeasureSpec);
        setMeasuredDimension(mMeasureHelper.getMeasuredWidth(), mMeasureHelper.getMeasuredHeight());
    }


}

File: dkplayer-java/src/main/java/xyz/doikki/videoplayer/render/TextureViewRender.kt
package xyz.doikki.videoplayer.render

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.SurfaceTexture
import android.util.AttributeSet
import android.view.Surface
import android.view.TextureView
import android.view.View
import xyz.doikki.videoplayer.GlobalConfig
import xyz.doikki.videoplayer.player.IPlayer
import xyz.doikki.videoplayer.render.Render.Companion.createShotBitmap
import xyz.doikki.videoplayer.render.Render.ScreenShotCallback
import xyz.doikki.videoplayer.render.Render.SurfaceListener
import xyz.doikki.videoplayer.util.L
import java.lang.ref.WeakReference

/**
 * 经查看[TextureView]源码，发现在[.onDetachedFromWindow]的时候，
 * 会先回调[TextureView.SurfaceTextureListener.onSurfaceTextureDestroyed]并释放[SurfaceTexture],
 * 在需要[SurfaceTexture]时会重新构建并回调[TextureView.SurfaceTextureListener.onSurfaceTextureAvailable]
 *
 * @see Render 具体可调用的方法请查看Render
 */
@SuppressLint("ViewConstructor")
class TextureViewRender : TextureView, Render {

    constructor(context: Context) : super(context)
    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs)
    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(
        context,
        attrs,
        defStyleAttr
    )

    private val proxy: RenderViewProxy = RenderViewProxy.new(this)
    private var playerRef: WeakReference<IPlayer>? = null
    private var st: SurfaceTexture? = null
    private var surface: Surface? = null
    private var surfaceListener: SurfaceListener? = null

    private val isEnableRenderOptimization: Boolean = GlobalConfig.isTextureViewRenderOptimizationEnabled

    private val stListener: SurfaceTextureListener = object : SurfaceTextureListener {

        override fun onSurfaceTextureAvailable(
            surface: SurfaceTexture,
            width: Int,
            height: Int
        ) {
            L.d("TextureViewRender onSurfaceTextureAvailable $st $width $height ${playerRef?.get()}")
            if (isEnableRenderOptimization) {
                //开启渲染优化
                if (st == null) {
                    st = surfaceTexture
                    this@TextureViewRender.surface = Surface(st)
                    bindSurfaceToMediaPlayer(this@TextureViewRender.surface!!)
                } else {
                    //在开启优化的情况下，使用最开始的那个渲染器
                    setSurfaceTexture(st!!)
                }
            } else {
                this@TextureViewRender.surface = Surface(st)
                bindSurfaceToMediaPlayer(this@TextureViewRender.surface!!)
            }
            notifySurfaceAvailable(this@TextureViewRender.surface, width, height)
        }

        override fun onSurfaceTextureSizeChanged(
            surface: SurfaceTexture,
            width: Int,
            height: Int
        ) {
            L.d("TextureViewRender onSurfaceTextureSizeChanged $st $width $height")
            surfaceListener?.onSurfaceSizeChanged(this@TextureViewRender.surface, width, height)
        }

        override fun onSurfaceTextureDestroyed(surface: SurfaceTexture): Boolean {
            L.d("TextureViewRender onSurfaceTextureDestroyed $st")
            //清空释放
            surfaceListener?.onSurfaceDestroyed(this@TextureViewRender.surface)
            return if (isEnableRenderOptimization) {
                //如果开启了渲染优化，那mSurfaceTexture通常情况不可能为null（在onSurfaceTextureAvailable初次回调的时候被赋值了），
                // 所以这里通常返回的是false，返回值false会告诉父类不要释放SurfaceTexture
                st == null
            } else {
                true
            }
        }

        override fun onSurfaceTextureUpdated(surface: SurfaceTexture) {
            if (isEnableRenderOptimization && st == null) {
                //用于修正开启渲染优化的情况下，mSurfaceTexture为空的异常情况：据说是因为存在机型不回调onSurfaceTextureAvailable而只回调此方法
                st = surface
                this@TextureViewRender.surface = Surface(surface)
                bindSurfaceToMediaPlayer(this@TextureViewRender.surface!!)
                notifySurfaceAvailable(this@TextureViewRender.surface, width, height)
            } else if (this@TextureViewRender.surface == null) {
                //用于修正未开启渲染优化的情况下，mSurface为空的异常情况：据说是因为存在机型不回调onSurfaceTextureAvailable而只回调此方法
                this@TextureViewRender.surface = Surface(surfaceTexture)
                bindSurfaceToMediaPlayer(this@TextureViewRender.surface!!)
                notifySurfaceAvailable(this@TextureViewRender.surface, width, height)
            } else {
                //onSurfaceTextureUpdated会在SurfaceTexture.updateTexImage()的时候回调该方法，因此只要图像更新，都会调用本方法，因此在这个方法中不适合做什么处理
//                Log.d("TextureView", "onSurfaceTextureUpdated $surface")
//                mSurfaceListener?.onSurfaceUpdated(mSurface)
            }
        }

        private fun bindSurfaceToMediaPlayer(surface: Surface) {
            playerRef?.get()?.setSurface(surface)
        }

        private fun notifySurfaceAvailable(surface: Surface?, width: Int, height: Int) {
            surfaceListener?.onSurfaceAvailable(surface)
        }
    }

    /**
     * 绑定播放器
     *
     * @param player
     */
    override fun attachPlayer(player: IPlayer) {
        playerRef = WeakReference(player)
        //当前surface不为空，则说明是surface重用
        surface?.let {
            player.setSurface(it)
        }
    }

    override fun setVideoSize(videoWidth: Int, videoHeight: Int) {
        proxy.setVideoSize(videoWidth, videoHeight)
    }

    override val view: View = this

    override fun setSurfaceListener(listener: SurfaceListener?) {
        surfaceListener = listener
    }

    override fun screenshot(highQuality: Boolean, callback: ScreenShotCallback) {
        if (!isAvailable) {
            callback.onScreenShotResult(null)
            return
        }
        if (highQuality) {
            callback.onScreenShotResult(bitmap)
        } else {
            callback.onScreenShotResult(getBitmap(createShotBitmap(this, false)))
        }
    }

    override fun setAspectRatioType(aspectRatioType: Int) {
        proxy.setAspectRatioType(aspectRatioType)
    }

    override fun setVideoRotation(degree: Int) {
        proxy.setVideoRotation(degree)
    }

    override fun setMirrorRotation(enable: Boolean) {
        proxy.setMirrorRotation(enable)
    }

    override fun release() {
        surface?.release()
        st?.release()
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        proxy.doMeasure(widthMeasureSpec, heightMeasureSpec)
        setMeasuredDimension(proxy.measuredWidth, proxy.measuredHeight)
    }

    init {
        surfaceTextureListener = stListener
    }
}